util/types/field_type.go

```diff
@@ -91,7 +91,7 @@ func AggTypeClass(tps []*FieldType, flag *uint) TypeClass {
 			continue
 		}
 		argTypeClass := argFieldType.ToClass()
-		if argTypeClass == ClassString && mysql.HasBinaryFlag(argFieldType.Flag) {
+		if (IsTypeBlob(argFieldType.Tp) || IsTypeVarchar(argFieldType.Tp) || IsTypeChar(argFieldType.Tp)) && mysql.HasBinaryFlag(argFieldType.Flag) {
```

Comment:
@XuHuaiyu I found some other inconsistencies for control-builtin due to `charset.CollationBin` and `mysql.BinaryFlag` are not set together. For example:
**In mysql:**
```
mysql> select ifnull(null,null);
Field   1:  `ifnull(null,null)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NULL
Collation:  binary (63)
Length:     0
Max_length: 0
Decimals:   31
Flags:      BINARY NUM

mysql> select case when null  then null else null end;
Field   1:  `case when null  then null else null end`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NULL
Collation:  binary (63)
Length:     0
Max_length: 0
Decimals:   31
Flags:      BINARY NUM 


```
**In tidb:**
```
tidb> select ifnull(null,null);
Field   1:  `ifnull(null,null)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NULL
Collation:  utf8_general_ci (33)
Length:     0
Max_length: 0
Decimals:   0
Flags:      BINARY NUM

tidb> select case when null  then null else null end;
Field   1:  `case when null  then null else null end`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       NULL
Collation:  utf8_general_ci (33)
Length:     0
Max_length: 0
Decimals:   0
Flags:      BINARY NUM 

```

Maybe we should rewrite AggTypeClass, put all `Flag`, `Charset` matters together